src/db.rs

```diff
@@ -354,1054 +256,442 @@ impl DB {
-            let start = start.as_ref().map(AsRef::as_ref);
-            let end = end.as_ref().map(AsRef::as_ref);
-
-            ffi::rocksdb_compact_range_cf_opt(
-                self.inner,
-                cf.inner,
-                opts.inner,
-                opt_bytes_to_ptr(start),
-                start.map_or(0, |s| s.len()) as size_t,
-                opt_bytes_to_ptr(end),
-                end.map_or(0, |e| e.len()) as size_t,
-            );
+            ffi_try!(ffi::rocksdb_destroy_db(opts.inner, cpath.as_ptr()));
         }
+        Ok(())
     }
 
-    pub fn set_options(&self, opts: &[(&str, &str)]) -> Result<(), Error> {
-        let copts = opts
-            .iter()
-            .map(|(name, value)| {
-                let cname = match CString::new(name.as_bytes()) {
-                    Ok(cname) => cname,
-                    Err(e) => return Err(Error::new(format!("Invalid option name `{}`", e))),
-                };
-                let cvalue = match CString::new(value.as_bytes()) {
-                    Ok(cvalue) => cvalue,
-                    Err(e) => return Err(Error::new(format!("Invalid option value: `{}`", e))),
-                };
-                Ok((cname, cvalue))
-            })
-            .collect::<Result<Vec<(CString, CString)>, Error>>()?;
-
-        let cnames: Vec<*const c_char> = copts.iter().map(|opt| opt.0.as_ptr()).collect();
-        let cvalues: Vec<*const c_char> = copts.iter().map(|opt| opt.1.as_ptr()).collect();
-        let count = opts.len() as i32;
+    pub fn repair<P: AsRef<Path>>(opts: &Options, path: P) -> Result<(), Error> {
+        let cpath = to_cpath(path)?;
         unsafe {
-            ffi_try!(ffi::rocksdb_set_options(
-                self.inner,
-                count,
-                cnames.as_ptr(),
-                cvalues.as_ptr(),
-            ));
+            ffi_try!(ffi::rocksdb_repair_db(opts.inner, cpath.as_ptr()));
         }
         Ok(())
     }
+}
 
-    /// Retrieves a RocksDB property by name.
-    ///
-    /// Full list of properties could be find
-    /// [here](https://github.com/facebook/rocksdb/blob/08809f5e6cd9cc4bc3958dd4d59457ae78c76660/include/rocksdb/db.h#L428-L634).
-    pub fn property_value(&self, name: &str) -> Result<Option<String>, Error> {
-        let prop_name = match CString::new(name) {
-            Ok(c) => c,
-            Err(e) => {
-                return Err(Error::new(format!(
-                    "Failed to convert property name to CString: {}",
-                    e
-                )));
-            }
-        };
-
-        unsafe {
-            let value = ffi::rocksdb_property_value(self.inner, prop_name.as_ptr());
-            if value.is_null() {
-                return Ok(None);
+macro_rules! make_new_db_with_traits {
+    ($struct_name:ident, [$($d:ty),+])  => (
+        #[derive(Delegate)]
+        $(#[delegate($d)])+
+        pub struct $struct_name(DBInner);
+
+        impl $struct_name {
+            delegate! {
```

Comment:
I'm not sure that the delegation is that useful herel. Personally, I would replace this delegation with a manual implementation, it wouldn't require much more boilerplate, right?

Reply:
Yes, not that much, each method will expand to 4 lines, for example:
`pub fn path(&self) -> & Path;`  expands to
```
#[always(inline)]
pub fn path(&self) -> & Path {
  self.0.path()
}
```

Reply:
If we know that the compiler is already inlining this in release mode we can remove the attribute.